src/Compilers/Core/Portable/ExtendedSpecialType.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A structure meant to represent a union of <see cref="SpecialType"/> and <see cref="InternalSpecialType"/> values
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit, Size = 1)]
+    internal readonly struct ExtendedSpecialType
+    {
+        [FieldOffset(0)]
+        private readonly sbyte _value;
+
+        private ExtendedSpecialType(int value)
+        {
+            Debug.Assert(value >= sbyte.MinValue && value <= sbyte.MaxValue);
```

Comment:
Consider asserting that `Debug.Assert(value < (int)InternalSpecialType.NextAvailable);`

Reply:
> Consider asserting that `Debug.Assert(value < (int)InternalSpecialType.NextAvailable);`

It is not my goal to enforce that values correspond to the actual set of elements defined in the enums. Like with regular enum type, one can assign any value that fits into the underlying type